core/src/main/java/feast/core/service/SpecService.java

```diff
@@ -206,26 +235,46 @@ public ListStoresResponse listStores(ListStoresRequest.Filter filter) {
   }
 
   /**
-   * Adds the featureSet to the repository, and prepares the sink for the feature creator to write
-   * to. If there is a change in the featureSet's schema or source, the featureSet version will be
-   * incremented.
+   * Creates or updates a feature set in the repository. If there is a change in the feature set
+   * schema, then the feature set version will be incremented.
    *
-   * <p>This function is idempotent. If no changes are detected in the incoming featureSet's schema,
-   * this method will update the incoming featureSet spec with the latest version stored in the
-   * repository, and return that.
+   * <p>This function is idempotent. If no changes are detected in the incoming featureSet's
+   * schema, this method will update the incoming featureSet spec with the latest version stored in
+   * the repository, and return that.
    *
-   * @param newFeatureSet featureSet to add.
+   * @param newFeatureSet Feature set that will be created or updated.
    */
-  public ApplyFeatureSetResponse applyFeatureSet(FeatureSetProto.FeatureSet newFeatureSet)
-      throws InvalidProtocolBufferException {
+  public ApplyFeatureSetResponse applyFeatureSet(FeatureSetProto.FeatureSet newFeatureSet) {
 
+    // Validate incoming feature set
     FeatureSetValidator.validateSpec(newFeatureSet);
+
+    // Ensure that the project already exists
+    String project_name = newFeatureSet.getSpec().getProject();
+    Project project = projectRepository.findById(newFeatureSet.getSpec().getProject())
+        .orElseThrow(() -> new IllegalArgumentException(String
+            .format("Project name does not exist. Please create a project first: %s", project_name
+            )));
+
+    // Ensure that the project is not archived
+    if (project.isArchived()) {
+      throw new IllegalArgumentException(String
```

Comment:
No way to retrieve featuresets under archived projects?

Reply:
Not yet. I think we can consider this as unimplemented. Should be easy to add later.

Reply:
ok